lib/operations/crud_write.js

```diff
@@ -0,0 +1,284 @@
+
+  // Return the results
+  handleCallback(callback, null, result);
+}
+
+class InsertOneModel extends BulkInsertModel {
+  constructor(args) {
+    // TODO: Destructure
+    const collection = args.collection;
+    const doc = args.doc;
+    const options = args.options;
+
+    super(collection, options);
+    if (this.options.keepGoing) {
+      this.options.ordered = false;
+    }
+
+    this.operations = prepareDocs(collection, [doc], options);
+  }
+
+  execute(server, callback) {
+    return super.execute(server, (err, r) => insertOneCallback(err, r, this.operations, callback));
+  }
+}
+
+function updateCallback(err, r, callback) {
+  if (callback == null) return;
+  if (err) return callback(err);
+  if (r == null) return callback(null, { result: { ok: 1 } });
+  r.modifiedCount = r.result.nModified != null ? r.result.nModified : r.result.n;
+  r.upsertedId =
+    Array.isArray(r.result.upserted) && r.result.upserted.length > 0
+      ? r.result.upserted[0] // FIXME(major): should be `r.result.upserted[0]._id`
+      : null;
+  r.upsertedCount =
+    Array.isArray(r.result.upserted) && r.result.upserted.length ? r.result.upserted.length : 0;
+  r.matchedCount =
+    Array.isArray(r.result.upserted) && r.result.upserted.length > 0 ? 0 : r.result.n;
+  callback(null, r);
+}
+
+class BulkUpdateModel extends CrudModel {
+  constructor(collection, options) {
+    super(collection, options);
+    this.type = UPDATE;
+  }
+
+  canRetry() {
+    return !this.operations.some(op => op.multi);
+  }
+}
+
+class UpdateModel extends BulkUpdateModel {
+  constructor(args) {
+    // TODO: Destructure
+    const collection = args.collection;
+    const selector = args.selector;
+    const update = args.update;
+    const options = args.options;
+    const multi = args.multi;
+
+    super(collection, options);
+    this.options.multi = multi;
+
+    const op = { q: selector, u: update };
+
+    op.upsert = this.options.upsert !== void 0 ? !!this.options.upsert : false;
+    op.multi = this.options.multi !== void 0 ? !!this.options.multi : false;
+
+    if (options.arrayFilters) {
+      op.arrayFilters = options.arrayFilters;
+      delete this.options.arrayFilters;
+    }
+
+    decorateWithCollation(this.options, collection, options);
+    this.operations = [op];
+  }
+
+  execute(server, callback) {
+    return super.execute(server, (err, r) => updateCallback(err, r, callback));
+  }
+}
+
+class ReplaceModel extends UpdateModel {
+  execute(server, callback) {
+    return super.execute(server, (err, r) => {
+      if (err) return callback(err, null);
+      r.ops = this.operations;
+      callback(null, r);
+    });
+  }
+}
+
+function deleteCallback(err, r, callback) {
+  if (callback == null) return;
+  if (err && callback) return callback(err);
+  if (r == null) return callback(null, { result: { ok: 1 } });
+  r.deletedCount = r.result.n;
+  if (callback) callback(null, r);
+}
+
+class BulkRemoveModel extends CrudModel {
```

Comment:
there is a `Bulk*Model` for each of these models, but they aren't subclasses by anything else. Do we need this much hierarchy? Are you anticipating needing these in the future? I'd say KISS here.

Reply:
I'm using them in #2002 .

Reply:
I don't find that answer compelling 😄 If we're doing this work to make more sense of the whole way we make write operations, we should be able to come up with a solution that doesn't require the deep hierarchy represented here. Please take a look at e.g. [Java's operations](https://github.com/mongodb/mongo-java-driver/blob/master/driver-core/src/main/com/mongodb/internal/operation/Operations.java#L307) to get a better sense of how this can be accomplished without this complexity:

Reply:
The key behind each Bulk model is that they declare how to detect if the operation is retryable. With those bases, we get to keep the logic in one place for each type of CRUD op, instead of spreading it out across multiple files like it previously was.